python_modules/libraries/dagster-dbt/dagster_dbt/dbt_artifacts.py

```diff
@@ -0,0 +1,134 @@
+        * During development, reload the manifest during start-up to pick up any changes.
+        * When deploying, use a compiled manifest to reduce start-up time.
+        * Handle the scenario when the dbt project is copied in to a directory for packaging.
+
+        Args:
+            project_dir (Union[str, Path]):
+                The directory of the dbt project.
+            target_folder (Union[str, Path]):
+                The folder in the project project directory to output artifacts.
+                Default: "target"
+            prepare_command: The dbt cli command to run to prepare the manifest.json
+                Default: ["parse", "--quiet"]
+
+        """
+        self.project_dir = project_dir if isinstance(project_dir, Path) else Path(project_dir)
+        self._target_folder = (
+            target_folder if isinstance(target_folder, Path) else Path(target_folder)
+        )
+        self._prepare_command = prepare_command
+
+        if package_data_dir is None:
+            self._package_data_dir = None
+        elif isinstance(package_data_dir, Path):
+            self._package_data_dir = package_data_dir
+        else:
+            self._package_data_dir = Path(package_data_dir)
+
+    @cached_property  # to prevent repeat compilation
+    def manifest_path(self) -> Path:
+        """The path to the manifest.json, compiling the manifest first if in development or
+        ensuring it already exists if not.
+        """
+        if _should_compile():
+            return self._prepare_manifest()
+
+        p = self._base_dir / self._target_folder / "manifest.json"
+        if not p.exists():
+            raise DagsterDbtManifestNotPreparedError(
+                f"Did not find prepared manifest.json at expected path {p}.\n"
+                "If this is an environment that is expected to prepare the manifest on load, "
+                "set the environment variable DAGSTER_DBT_PARSE_PROJECT_ON_LOAD."
+            )
+        return p
+
+    def prepare(self) -> None:
+        """A method that can be called as part of the deployment process to handle
+        preparing the manifest and if package_data_dir is set, handle copying
+        the dbt project to that directory.
+        """
+        self._prepare_manifest()
+        self._handle_package_data()
+
+    def get_cli_resource(self, **kwargs) -> DbtCliResource:
+        """Construct a DbtCliResource that targets the correct project directory.
+        If package_data_dir is set, resolves to that outside of the development context.
+        """
+        if _should_compile():
+            return DbtCliResource(project_dir=os.fspath(self.project_dir))
+
+        return DbtCliResource(
+            project_dir=os.fspath(self._base_dir),
+            **kwargs,
+        )
+
+    @property
+    def _base_dir(self) -> Path:
+        return self._package_data_dir if self._package_data_dir else self.project_dir
+
+    def _prepare_manifest(self) -> Path:
```

Comment:
I can imagine that this is an abstractmethod for overriding this class in the future

Reply:
Thinking was that `prepare_command` would avoid a good chunk of cases. Do you think `_` prefix should get dropped if people might override?